file.go

```diff
@@ -7,56 +7,109 @@ import (
 )
 
 var (
-	ErrNotDirectory = errors.New("Couldn't call NextFile(), this isn't a directory")
-	ErrNotReader    = errors.New("This file is a directory, can't use Reader functions")
+	ErrNotDirectory = errors.New("file isn't a directory")
+	ErrNotReader    = errors.New("file isn't a regular file")
+
+	ErrNotSupported = errors.New("operation not supported")
 )
 
 // File is an interface that provides functionality for handling
 // files/directories as values that can be supplied to commands. For
-// directories, child files are accessed serially by calling `NextFile()`.
+// directories, child files are accessed serially by calling `NextFile()`
+//
+// Read/Seek methods are only valid for files
+// NextFile method is only valid for directories
 type File interface {
-	// Files implement ReadCloser, but can only be read from or closed if
-	// they are not directories
-	io.ReadCloser
-
-	// FileName returns a filename associated with this file
-	FileName() string
-
-	// FullPath returns the full path used when adding with this file
-	FullPath() string
-
-	// IsDirectory returns true if the File is a directory (and therefore
-	// supports calling `NextFile`) and false if the File is a normal file
-	// (and therefor supports calling `Read` and `Close`)
-	IsDirectory() bool
-
-	// NextFile returns the next child file available (if the File is a
-	// directory). It will return (nil, io.EOF) if no more files are
-	// available. If the file is a regular file (not a directory), NextFile
-	// will return a non-nil error.
-	NextFile() (File, error)
+	io.Closer
+
+	// Size returns size of this file (if this file is a directory, total size of
+	// all files stored in the tree should be returned). Some implementations may
+	// choose not to implement this
+	Size() (int64, error)
 }
 
-type StatFile interface {
+// Regular represents a regular Unix file
+type Regular interface {
 	File
 
-	Stat() os.FileInfo
+	io.Reader
+	io.Seeker
+}
+
+// DirEntry exposes information about a directory entry
+type DirEntry interface {
+	// Name returns the base name of this entry, which is the base name of
+	// the referenced file
+	Name() string
+
+	// File returns the file referenced by this DirEntry
+	File() File
+
+	// Regular is an alias for ent.File().(Regular). If the file isn't a regular
+	// file, nil value will be returned
+	Regular() Regular
```

Comment:
So I'm not a fan of making this api too big, but:
* I don't think we will be getting any major new file types as common as directory/file any time soon
* Those are just 2 types (+the base type for other stuff such as casting to symlink)
* It's so much nicer to use this vs casting when we know what file type we expect I think it might be worth keeping

Reply:
For now there are two types. In the future we might have more, for example`Symlink`, etc might become a distinct type.  @Stebalien thoughts?

Reply:
I simply wouldn't add stuff like `Symlink` here, those 2 are just very convenient helpers for 2 most common cases. If you think this is making things too inconsistent, I'm fine with removing this, but I'd rather keep it based on how much nicer the api is with those helpers

Reply:
I have a slight preference to keeping the interface minimal, but not a strong one.  It just seams like the type of think we should be using a helper function for, even though it is a little less convenient to use.

Reply:
@Stebalien opinions?

Reply:
I see that these methods have been removed. I assume the new way to do this is just to type assert?

However, responding to the question: my only concern with having each of these methods is that, to me at least, they imply that there are no other options.

Really, we can probably just have 4: Special, Symlink, Regular, and Directory (where special is a catch-all) but it's probably simpler to just do a type switch. Is it cleaner to try calling each of these methods than to do a type assert (assuming you have to check all cases anyways)?

Reply:
There are also `ToFile`/`ToDir` helpers which basically do the cast or return nil, and `FileFrom`/`DirFrom` which are convenience functions for getting those types from various things like iterators or raw readers.

I found this seems to work the best as on the api consumer level we usually care about one of the following:
* We want to assert that some `Node` is a `File/Directory/Whatever` (doable with `To*` functions)
* We want to easily switch over types (we use interfaces so it's easy enough)
  * No, calling each method isn't the nicest option unless there is a lot of shared code between cases
* We want to easily construct/extract files from Iterators, raw `Reader`s, directories from`map[string]Node`, etc. (doable with `*From` functions)
  * `*From` replaced cast functions on `DirEntry`

Reply:
Commented inline.